src/wallet/coinselection.cpp

```diff
@@ -0,0 +1,349 @@
+        // while((i != aEnd) && ((i->first + grp.first < targetValue) || (grp.second.find(i) != grp.second.end()))  )
+        // allow dups and eliminate later for performance
+        while ((i != aEnd) && (i->first + grp.first < targetValue)) // iterate to make sure it is actually big enough.
+        {
+            ++i;
+        }
+
+        if (i != aEnd) // Ok we have a solution so add it.
+        {
+            TxoGroup g = makeGroup(i, &grp);
+            solutions.insert(TxoGroupMap::value_type(g.first, g));
+            ret = true;
+        }
+    }
+
+    if (small == aEnd)
+        --small;
+    for (int i = 0; i < 5; i++) // check 5 more close values
+    {
+        // lower_bound returns an element >= the passed value, so get one that is smaller
+        while ((small != aBeg) && (small->first + grp.first > targetValue))
+            --small;
+        // keep looking if there are no solutions or if the depth is low
+        if ((solutions.empty() || (depth < MAX_ELECTIVE_TXOS)))
+        {
+            // Its better performance to accidentally add a double than to check every time
+            // go to the prev if this one is already in the group.
+            // while ((small != aBeg) && (grp.second.find(small) != grp.second.end())) --small;
+            // if (small != aBeg)
+            {
+                TxoGroup newGrp = makeGroup(small, &grp);
+                ret |= extendCoinSelectionSolution(targetValue, available, newGrp, solutions, depth + 1);
+            }
+        }
+
+        if (small == aBeg)
+            break;
+    }
+    return ret;
+}
+
+// Make sure that the group sums to >= the target value.
+// Since the TxoGroup is a set, identical Txos are eliminated meaning that this may not sum, and that the
+// TxoGroup->first may be inaccurate.
+// There may be identical Txos because the search algorithm does not eliminate used txos from the search set for
+// efficiency.
+// Also check various other aspects that would block the transaction from being accepted into the mempool.
+bool validate(const TxoGroup &grp, const CAmount targetValue)
+{
+    CAmount acc = 0;
+    int nTx = grp.second.size();
+    for (TxoItVec::iterator i = grp.second.begin(); i != grp.second.end(); ++i)
+    {
+        acc += (*i)->first;
+    }
+    if (acc < targetValue)
+        return false;
+
+    std::vector<CTxIn> txIn(nTx);
+    int count = 0;
+    for (TxoItVec::iterator i = grp.second.begin(); i != grp.second.end(); ++i, ++count)
+    {
+        COutput &tmp = (*i)->second;
+        txIn[count] = CTxIn(tmp.tx->GetHash(), tmp.i);
+    }
+
+    size_t limitAncestorCount = GetArg("-limitancestorcount", BU_DEFAULT_ANCESTOR_LIMIT);
+    size_t limitAncestorSize = GetArg("-limitancestorsize", BU_DEFAULT_ANCESTOR_SIZE_LIMIT) * 1000;
+    size_t limitDescendantsCount = GetArg("-limitdescendantcount", BU_DEFAULT_DESCENDANT_LIMIT);
+    size_t limitDescendantSize = GetArg("-limitdescendantsize", BU_DEFAULT_DESCENDANT_SIZE_LIMIT) * 1000;
```

Comment:
we have this snippet of code that declare and init this 4 set of var (with slightly different names) also in `src/txadmission.cpp` (L1127, ParallelAcceptToMemoryPool()`) and in `src/net_processing.cpp` (L591, ProcessMessage()`). 

I wonder if we can consolidate the code somehow.

Reply:
We could make them globals.  Given our work and planned experiments perhaps it makes sense to make them tweaks?

Reply:
Having tweaks for those would be very useful

Reply:
let's do it in a separate PR?

Reply:
ok by me